File: cli/src/fp_cli/commands/logs.py
"""fp logs — tail app logs."""

import json
import re

import typer

from ..api.client import PlatformClient, PlatformError, resolve_app
from ..config import get_active_platform_or_exit
from ..console import console, err_console
from ..project import find_project_file


def _resolve_app_name(name: str | None) -> str:
    if name:
        return name
    fp_file = find_project_file()
    if fp_file:
        import yaml

        data = yaml.safe_load(fp_file.read_text()) or {}
        if data.get("name"):
            return data["name"]
    err_console.print("[red]No app name provided and no .fp.yaml found.[/red]")
    raise typer.Exit(1)


def _parse_since(since: str) -> int:
    """Parse duration string like '1h', '30m', '5s' to seconds."""
    match = re.match(r"^(\d+)([smh])$", since.strip())
    if not match:
        err_console.print(f"[red]Invalid --since format: '{since}'.[/red] Use e.g. 30s, 5m, 1h")
        raise typer.Exit(1)
    value, unit = int(match.group(1)), match.group(2)
    multiplier = {"s": 1, "m": 60, "h": 3600}
    return value * multiplier[unit]


def logs(
    name: str = typer.Argument(None, help="App name (uses .fp.yaml if omitted)"),
    follow: bool = typer.Option(True, "--follow/--no-follow", "-f/-F", help="Stream logs in real-time"),
    since: str = typer.Option("", "--since", "-s", help="Only logs since (e.g. 30s, 5m, 1h)"),
    tail: int = typer.Option(100, "--tail", "-n", help="Number of recent lines"),
):
    """Tail app logs."""
    app_name = _resolve_app_name(name)
    platform = get_active_platform_or_exit()
    client = PlatformClient(platform["url"], platform["token"])

    app = resolve_app(client, app_name)
    app_id = app["app_id"]

    if follow and not since:
        _stream_logs_ws(platform, app_id)
    else:
        _fetch_logs_http(client, app_id, tail, since)


def _fetch_logs_http(client: PlatformClient, app_id: str, tail: int, since: str):
    """Fetch logs via HTTP and print."""
    since_seconds = _parse_since(since) if since else None

    try:
        data = client.get_logs(app_id, tail_lines=tail, since_seconds=since_seconds)
    except PlatformError as e:
        err_console.print(f"[red]Failed to fetch logs:[/red] {e.message}")
        raise typer.Exit(1)

    log_lines = data.get("logs", [])
    if not log_lines:
        console.print("[dim]No logs available.[/dim]")
        return

    for line in log_lines:
        ts = line.get("timestamp", "")
        msg = line.get("message", "")
        if ts:
            console.print(f"[dim]{ts}[/dim] {msg}")
        else:
            console.print(msg)


def _stream_logs_ws(platform: dict, app_id: str):
    """Stream logs via WebSocket."""
    import websockets.sync.client as ws_client

    # Build WebSocket URL from platform URL
    ws_url = platform["url"].replace("https://", "wss://").replace("http://", "ws://")
    url = f"{ws_url}/api/apps/{app_id}/logs/stream?token={platform['token']}"

    console.print(f"[dim]Streaming logs (Ctrl+C to stop)...[/dim]")

    try:
        with ws_client.connect(url) as ws:
            for message in ws:
                try:
                    data = json.loads(message)
                except (json.JSONDecodeError, TypeError):
                    console.print(str(message))
                    continue

                msg_type = data.get("type", "")
                if msg_type == "log":
                    ts = data.get("timestamp", "")
                    msg = data.get("message", "")
                    if ts:
                        console.print(f"[dim]{ts}[/dim] {msg}")
                    else:
                        console.print(msg)
                elif msg_type == "connected":
                    console.print(f"[dim]Connected to pod: {data.get('pod_name', 'unknown')}[/dim]")
                elif msg_type == "status":
                    console.print(f"[yellow]{data.get('message', '')}[/yellow]")
                elif msg_type == "error":
                    err_console.print(f"[red]{data.get('message', '')}[/red]")
    except KeyboardInterrupt:
        console.print("\n[dim]Stopped.[/dim]")
    except Exception as e:
        err_console.print(f"[red]WebSocket error:[/red] {e}")
        console.print("[dim]Falling back to HTTP logs...[/dim]")
        # Fallback to HTTP
        from ..api.client import PlatformClient

        client = PlatformClient(platform["url"], platform["token"])
        _fetch_logs_http(client, app_id, 50, "")

File: cli-go/cmd/logs.go
package cmd

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/url"
	"os"
	"regexp"
	"strings"

	"github.com/gorilla/websocket"
	"github.com/spf13/cobra"
	"github.com/thatcatxedo/fastapi-platform/cli-go/internal/api"
	"github.com/thatcatxedo/fastapi-platform/cli-go/internal/config"
)

var logsCmd = &cobra.Command{
	Use:   "logs [name]",
	Short: "Tail app logs",
	Args:  cobra.MaximumNArgs(1),
	RunE:  runLogs,
}

var (
	logsFollow   bool
	logsNoFollow bool
	logsSince    string
	logsTail     int
)

func init() {
	logsCmd.Flags().BoolVarP(&logsFollow, "follow", "f", true, "Stream logs in real-time")
	logsCmd.Flags().BoolVar(&logsNoFollow, "no-follow", false, "Fetch recent logs only (no stream)")
	logsCmd.Flags().StringVarP(&logsSince, "since", "s", "", "Only logs since (e.g. 30s, 5m, 1h)")
	logsCmd.Flags().IntVarP(&logsTail, "tail", "n", 100, "Number of recent lines")
}

func parseSince(s string) (int, error) {
	re := regexp.MustCompile(`^(\d+)([smh])$`)
	m := re.FindStringSubmatch(strings.TrimSpace(s))
	if m == nil {
		return 0, fmt.Errorf("invalid --since format: '%s'. Use e.g. 30s, 5m, 1h", s)
	}
	var value int
	fmt.Sscanf(m[1], "%d", &value)
	multiplier := map[string]int{"s": 1, "m": 60, "h": 3600}
	return value * multiplier[m[2]], nil
}

func runLogs(cmd *cobra.Command, args []string) error {
	appName := resolveAppName("")
	if len(args) > 0 {
		appName = args[0]
	}
	p := requirePlatform()
	client := api.NewClient(p.URL, p.Token)
	app, err := api.ResolveApp(client, appName)
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	}
	appID, _ := app["app_id"].(string)

	if logsFollow && !logsNoFollow && logsSince == "" {
		streamLogsWS(p.URL, p.Token, appID)
		return nil
	}

	var sinceSeconds *int
	if logsSince != "" {
		s, err := parseSince(logsSince)
		if err != nil {
			fmt.Fprintln(os.Stderr, err.Error())
			os.Exit(1)
		}
		sinceSeconds = &s
	}

	data, err := client.GetLogs(appID, logsTail, sinceSeconds)
	if err != nil {
		var pe *api.PlatformError
		if errors.As(err, &pe) {
			fmt.Fprintf(os.Stderr, "Failed to fetch logs: %s\n", pe.Message)
			os.Exit(1)
		}
		return err
	}

	logLines, _ := data["logs"].([]interface{})
	if len(logLines) == 0 {
		fmt.Println("No logs available.")
		return nil
	}

	for _, line := range logLines {
		m, _ := line.(map[string]interface{})
		ts, _ := m["timestamp"].(string)
		msg, _ := m["message"].(string)
		if ts != "" {
			fmt.Printf("%s %s\n", ts, msg)
		} else {
			fmt.Println(msg)
		}
	}
	return nil
}

func streamLogsWS(platformURL, token, appID string) {
	wsURL := strings.Replace(platformURL, "https://", "wss://", 1)
	wsURL = strings.Replace(wsURL, "http://", "ws://", 1)
	wsURL = wsURL + "/api/apps/" + appID + "/logs/stream?token=" + url.QueryEscape(token)

	fmt.Println("Streaming logs (Ctrl+C to stop)...")

	conn, _, err := websocket.DefaultDialer.Dial(wsURL, nil)
	if err != nil {
		fmt.Fprintf(os.Stderr, "WebSocket error: %v\n", err)
		fmt.Println("Falling back to HTTP logs...")
		p := config.GetActivePlatform()
		if p != nil {
			client := api.NewClient(p.URL, p.Token)
			data, _ := client.GetLogs(appID, 50, nil)
			if logs, _ := data["logs"].([]interface{}); len(logs) > 0 {
				for _, line := range logs {
					m, _ := line.(map[string]interface{})
					ts, _ := m["timestamp"].(string)
					msg, _ := m["message"].(string)
					if ts != "" {
						fmt.Printf("%s %s\n", ts, msg)
					} else {
						fmt.Println(msg)
					}
				}
			}
		}
		return
	}
	defer conn.Close()

	for {
		_, message, err := conn.ReadMessage()
		if err != nil {
			return
		}
		var data map[string]interface{}
		if json.Unmarshal(message, &data) != nil {
			fmt.Println(string(message))
			continue
		}
		msgType, _ := data["type"].(string)
		switch msgType {
		case "log":
			ts, _ := data["timestamp"].(string)
			msg, _ := data["message"].(string)
			if ts != "" {
				fmt.Printf("%s %s\n", ts, msg)
			} else {
				fmt.Println(msg)
			}
		case "connected":
			podName, _ := data["pod_name"].(string)
			fmt.Printf("Connected to pod: %s\n", podName)
		case "status":
			msg, _ := data["message"].(string)
			fmt.Println(msg)
		case "error":
			msg, _ := data["message"].(string)
			fmt.Fprintf(os.Stderr, "%s\n", msg)
		}
	}
}
